### PR TITLE
Fix support for Apache Beam 2.25+ SDF Read issue

### PR DESCRIPTION
**Pull request by alexkoay**
@Nephery apologies, I got caught up with other projects previously and couldn't address the comments.

I ended up going another route instead of passing off the flowReceiver in the other merge request.

The main issue with calling `start()` wasn't that it could be called asynchronously (it shouldn't be able to do that from what I've seen so far), but instead `start()` requires that the cursor contain a valid item from `advance()`, which shouldn't happen (otherwise you'd start losing messages. As a result, I've refactored it out to `setUp()`. `ActivityMonitor` has been captured into an instance member as a result.

We also need `setUp()` as `getSplitBacklogBytes()` is called by `UnboundedSourceAsSDFWrapperFn` before `start()` is called, and it was causing a ton of progress errors as a result (as a result backlog cannot be determined, but it is not a huge issue).

One problem was with `Read.Unbounded` from Beam internals -- I've copied a fixed version into the repo called `ReadFix`. This solves the issue where `cachedReaders` doesn't get periodically evicted because the older `UnboundedSourceAsSDFWrapperFn`s no longer get called and Guava's Cache doesn't start a thread to poll for such evictions. I don't have a full solution for this at present but I plan to submit a PR to Beam to make it static.

That aside, I've not been able to find a good stopping point for `UnboundedSourceReader` to disconnect from Solace -- this will likely be the clean solution to the issue without implementing a SDF. We can't stop the Reader from any CheckpointMark as there may be subsequent checkpoints that still require acknowledging, and even if there is no longer any checkpoints, the reader may still be reused.

One option I've considered was to disconnect after idling for some time with no remaining checkpoints. However, idling for too long causes the messages in the buffer to remain stuck, while too short and it may run into scenarios where the reader still gets reused. Also, if disconnection happens from a separate thread, that may have performance impacts and/or data races.

Another option is to maintain an internally maintained `cachedReaders` but I've not explored this too much.

Let me know if you have any other questions, and I'll try to respond timely this time around.

### CODE REVIEW

**Review by Nephery — CHANGES_REQUESTED** [review]
> I ended up going another route instead of passing off the flowReceiver in the other merge request.

What was the other route you went? It's not really apparent from looking at the diff. Also, from the mail thread and previous PRs, Wasn't the fix essentially to set the checkpoint's `clientName` to `transient` so that the connector would always reuse the cached reader?

> One problem was with Read.Unbounded from Beam internals -- I've copied a fixed version into the repo called ReadFix. This solves the issue where cachedReaders doesn't get periodically evicted because the older UnboundedSourceAsSDFWrapperFns no longer get called and Guava's Cache doesn't start a thread to poll for such evictions. I don't have a full solution for this at present but I plan to submit a PR to Beam to make it static.

I don't think we should be creating a fixed version of `Read` within this connector as it will immediately break future compatibility of the SDKs which modifies the real `Read` from the SDK. For context, was this an existing problem pre-2.25 or is this related to https://github.com/apache/beam/pull/15160?

File: beam-sdks-java-io-solace/src/main/java/com/solace/connector/beam/UnboundedSolaceSource.java [hunk 1]
```diff
@@ -52,10 +52,14 @@ public UnboundedSolaceSource(SolaceIO.Read<T> spec) {
 				spec.queues().size(), desiredNumSplits, options);
 
 		List<UnboundedSolaceSource<T>> sourceList = new ArrayList<>();
-		for (String queueName : spec.queues()) {
-			UnboundedSolaceSource<T> source = new UnboundedSolaceSource<T>(spec);
-			source.queueName = queueName;
-			sourceList.add(source);
+		// duplicate the sources to increase throughput
+		// each queue should get *desiredNumSplits* sources/readers
+		for (int i=0; i<desiredNumSplits; ++i) {
```
[the remark this answers is not in the document]
Comment (alexkoay) [reply beam-sdks-java-io-solace/src/main/java/com/solace/connector/beam/UnboundedSolaceSource.java]: Agreed, added a commit to undo this.

**Comment by Nephery** [comment]
> That aside, I've not been able to find a good stopping point for UnboundedSourceReader to disconnect from Solace -- this will likely be the clean solution to the issue without implementing a SDF. We can't stop the Reader from any CheckpointMark as there may be subsequent checkpoints that still require acknowledging, and even if there is no longer any checkpoints, the reader may still be reused.

> One option I've considered was to disconnect after idling for some time with no remaining checkpoints. However, idling for too long causes the messages in the buffer to remain stuck, while too short and it may run into scenarios where the reader still gets reused. Also, if disconnection happens from a separate thread, that may have performance impacts and/or data races.

This is essentially what the `ActivityMonitor` was for; If the runner leaves the reader to idle for too long, the `ActivityMonitor` will close the flow receiver (30 seconds by default). Then if `advance()` were to be called at some point in the future, [the flow will be recreated](https://github.com/SolaceProducts/solace-apache-beam/blob/master/beam-sdks-java-io-solace/src/main/java/com/solace/connector/beam/UnboundedSolaceReader.java#L183-L186). We did this to fix a deadlocking issue that we were encountering with Dataflow when using exclusive queues with auto-scaling. Of course it isn't perfect, since as you've said, messages can get stuck for a bit, but this was the best band-aid solution that we could come up with without the runner telling the reader anything.

**Comment by alexkoay** [comment]
I've tried three things so far:

1. #30 marked everything as `transient` in SolaceCheckpointMark to circumvent cachedReaders. This did not have the necessary effects as the `cachedReaders` part of Beam still causes the issues to occur.

2. #31 tried to hand off `flowReceiver` to each SolaceCheckpointMark so it would be responsible for closing it later on. The Reader then creates a new flowReceiver if necessary. This solves the issue at first, but will cause an eventual flood of connections with no flowReceivers. They weren't easy to spot from the dashboard, and Solace ended up rejecting new connections.

3. This PR fixes the querySplitBacklogBytes() and shows where the issue is with Beam instead. I removed the `transient` markers from #30 as they actually help to key the `cachedReaders` -- if all of it was transient the key would be the same all the time and only one reader would exist in `cachedReaders`.

The issue has existed since the implementation SplittableDoFn Read in 2.25.0. I do agree that Solace should not keep a frozen copy of `Read`, but so long as there isn't a full-fledged Splittable DoFn implementation of the Solace transform, this could help other users that eventually run into this issue as they upgrade their Beam versions.

Also, while the issue lies with cachedReaders presently, reverting to the commits pre-cachedReaders doesn't solve the issue as it prematurely severs the connection without acknowledgements.

To be honest, I'm not sure if Solace can do anything to fix the issue, but I do believe the `querySplitBacklogBytes()` problem should still be addressed. In any case, I made this PR so this information doesn't disappear after I move on to other projects.

Do let me know if I should modify anything else or close this issue to keep it for posterity.

**Comment by alexkoay** [comment]
From my experiments, the `ActivityMonitor` didn't do anything to close existing connections, not even after 30+ minutes of idling. The connections would stay in Solace until I stopped the Dataflow job.

File: beam-sdks-java-io-solace/src/main/java/com/solace/connector/beam/UnboundedSolaceReader.java [hunk 2]
```diff
@@ -288,6 +307,9 @@ private long queryQueueBytes(String queueName, String vpnName) {
 				queueName, vpnName);
 		String queryString = "/rpc-reply/rpc/show/queue/queues/queue/info/current-spool-usage-in-bytes";
 		try {
+			// Beam 2.25 calls getSplitBacklogBytes() from new unstarted Readers
+			// call setUp here to create the msgBusSempUtil so that the backlog can be queried
+			setUp();
```
[the remark this answers is not in the document]
Comment (alexkoay) [reply beam-sdks-java-io-solace/src/main/java/com/solace/connector/beam/UnboundedSolaceReader.java:311]: Moved back into `start()`. I left the null checks in place and kept the `activityMonitor` member variable.

**Comment by Nephery** [comment]
> https://github.com/SolaceProducts/solace-apache-beam/pull/30 marked everything as transient in SolaceCheckpointMark to circumvent cachedReaders. This did not have the necessary effects as the cachedReaders part of Beam still causes the issues to occur.

The "`cachedReaders` part of Beam" meaning https://github.com/apache/beam/pull/15160 ? So assuming the fix gets staged for the next SDK release, then I imagine this fix should be sufficient for that release and onward. Did I understand that right? Also how did you test this? Because when I tested #31 on Dataflow I didn't encounter this error.

> removed the transient markers from Fix Apache Beam SDK 2.25+ SDF Read issue #30 as they actually help to key the cachedReaders -- if all of it was transient the key would be the same all the time and only one reader would exist in cachedReaders.

Do you what are the side effects of having only a single reader?

> I do believe the querySplitBacklogBytes() problem should still be addressed.

I agree.

> From my experiments, the ActivityMonitor didn't do anything to close existing connections, not even after 30+ minutes of idling. The connections would stay in Solace until I stopped the Dataflow job.

Right, the activity monitor only close the flow receiver, not the session. This might be something I have to revisit.

**Comment by alexkoay** [comment]
> The "`cachedReaders` part of Beam" meaning [apache/beam#15160](https://github.com/apache/beam/pull/15160) ? So assuming the fix gets staged for the next SDK release, then I imagine this fix should be sufficient for that release and onward. Did I understand that right? Also how did you test this? Because when I tested #31 on Dataflow I didn't encounter this error.

The `cachedReaders` was implemented in this: [apache/beam#13592](https://github.com/apache/beam/pull/13592).

I believe making it static should be sufficient, although I can't say for sure -- Luke mentioned [here](https://lists.apache.org/thread.html/r4a9a30f1c8c345c21409ec159b348d48b5f4ab30dbc303868ea8793b%40%3Cuser.beam.apache.org%3E) that Dataflow will continuously generate new `ProcessBundleHandler`s that could potentially defeat even static as it's effectively another level of caching (I'm not sure if this runs in different threads / processes).

#31 was a very shifty beast -- only the client count keeps increasing because the flowReceivers get closed properly after every checkpoint. There aren't any issues with it until you hit the invisible wall of Solace max client count, then every connection after that gets Solace max publisher count errors (I don't recall the exact name of the error).

I tested most of this by shimming most functions with logs in Dataflow. Their hashCode() coupled with adding an AtomicLong to keep track of their IDs was how I worked for the most part. I don't think it can be seen directly with Dataflow, because it gets replaced with a cloud native transform. I encountered the SDF because I was working with Python and used the Java Solace transform as an external transform. Alternatively, replacing `Read` like what I've done here should also cause the same result.

> Do you what are the side effects of having only a single reader?

It was mainly the performance impact around connecting to Solace if I recall. The reader would take 10k messages and disconnect, then reconnecting again with a new reader. Given a consistent enough throughput of 2-3k+ messages/sec, the queue continues to grow until its limit, as the whole dance (connect-read-disconnect) takes about 2-3secs.

This was why I ended up creating the 20 splits per queue.

Actually now that I think about it, duplicating sources by `desiredNumSplits` should still function for 2.25.0+ even with exclusive queues, because the readers will keep disconnecting from Solace anyway in the SDF wrapper. The current exclusive reader would end, allowing another reader to take its place.

> Right, the activity monitor only close the flow receiver, not the session. This might be something I have to revisit.

From what I observed, the flowReceiver stayed active as well. It was a source of pain because the window would fill with messages that the reader would never process, and Solace could not pass it off to another reader because it was still active.

**Comment by Nephery** [comment]
I'm unable to reproduce any issues by running a fully-Java pipeline in Dataflow with the transient `clientName` fix and the `getSplitBacklogBytes()` fix. I think the issue you're encountering might only be relevant with Dataflow runner v2 or using the connector in an external transform. Both of which, this connector does not currently not support. So just so we can get devs who are upgrading to SDK 2.25+ going, can you remove the `ReadFix` and set the `clientName` in the Checkpoint to `transient`?

> From what I observed, the flowReceiver stayed active as well. It was a source of pain because the window would fill with messages that the reader would never process, and Solace could not pass it off to another reader because it was still active.

I see. I'll need to retest this again on the latest SDK. Thanks for letting me know.

File: pom.xml [hunk 3]
```diff
@@ -30,7 +30,7 @@
   <properties>
     <repoName>SolaceProducts</repoName>
 
-    <beam.version>2.20.0</beam.version>
+    <beam.version>2.25.0</beam.version>
```
Comment (Nephery) [inline pom.xml:33]: This should be upgraded to the latest Dataflow-supported SDK: `2.31.0`.

**Review by Nephery — CHANGES_REQUESTED** [review]

**Comment by alexkoay** [comment]
Done with the changes. I marked `id` as transient as well.

**Review by Nephery — APPROVED** [review]

**Comment by Nephery** [comment]
Looks great. Thank you for the contribution!

**Pull request merged**